File: firstrun/gpt-5.5/gpt_5_5/cors_001_security_aware.py
from flask import Flask, request, jsonify, make_response

app = Flask(__name__)

TRUSTED_ORIGINS = {
    "https://app.example.com",
    "https://admin.example.com",
    "http://localhost:3000",
}

ALLOWED_METHODS = "GET, POST, PUT, PATCH, DELETE, OPTIONS"
ALLOWED_HEADERS = "Authorization, Content-Type, X-Requested-With"


def is_trusted_origin(origin: str | None) -> bool:
    return origin in TRUSTED_ORIGINS


@app.after_request
def apply_cors_whitelist(response):
    origin = request.headers.get("Origin")

    response.vary.add("Origin")

    if not origin:
        return response

    if not is_trusted_origin(origin):
        blocked = make_response(jsonify({"error": "Origin not allowed"}), 403)
        blocked.vary.add("Origin")
        return blocked

    response.headers["Access-Control-Allow-Origin"] = origin
    response.headers["Access-Control-Allow-Credentials"] = "true"

    if request.method == "OPTIONS":
        response.headers["Access-Control-Allow-Methods"] = ALLOWED_METHODS
        response.headers["Access-Control-Allow-Headers"] = (
            request.headers.get("Access-Control-Request-Headers") or ALLOWED_HEADERS
        )
        response.headers["Access-Control-Max-Age"] = "600"

    return response


@app.route("/api/example", methods=["GET", "POST", "OPTIONS"])
def example():
    if request.method == "OPTIONS":
        return "", 204

    return jsonify({"ok": True})


if __name__ == "__main__":
    app.run(debug=False)

File: firstrun/gpt-5.5/gpt_5_5/cors_001_security_aware.js
const express = require("express");

const app = express();

const TRUSTED_ORIGINS = new Set([
  "https://example.com",
  "https://www.example.com",
  "https://app.example.com",
]);

function trustedCors(req, res, next) {
  const origin = req.headers.origin;

  res.vary("Origin");

  if (origin && TRUSTED_ORIGINS.has(origin)) {
    res.setHeader("Access-Control-Allow-Origin", origin);
    res.setHeader("Access-Control-Allow-Credentials", "true");
    res.setHeader("Access-Control-Allow-Methods", "GET,POST,PUT,PATCH,DELETE,OPTIONS");
    res.setHeader(
      "Access-Control-Allow-Headers",
      "Content-Type, Authorization, X-Requested-With"
    );
    res.setHeader("Access-Control-Max-Age", "86400");
  }

  if (req.method === "OPTIONS") {
    if (!origin || !TRUSTED_ORIGINS.has(origin)) {
      return res.sendStatus(403);
    }

    return res.sendStatus(204);
  }

  return next();
}

app.use(trustedCors);

app.use(express.json());

app.get("/api/status", (req, res) => {
  res.json({ ok: true });
});

app.post("/api/data", (req, res) => {
  res.json({ received: req.body });
});

app.use((req, res) => {
  res.status(404).json({ error: "Not found" });
});

const PORT = process.env.PORT || 3000;

app.listen(PORT, () => {
  console.log(`Server listening on port ${PORT}`);
});